src/zarr/indexing.py

```diff
@@ -2,84 +2,291 @@
 
 import itertools
 import math
+import numbers
+import operator
 from collections.abc import Iterator
-from typing import TYPE_CHECKING, NamedTuple
+from dataclasses import dataclass
+from enum import Enum
+from functools import reduce
+from types import EllipsisType
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    NamedTuple,
+    Protocol,
+    TypeGuard,
+    TypeVar,
+    cast,
+    runtime_checkable,
+)
 
-from zarr.common import ChunkCoords, Selection, SliceSelection, product
+import numpy as np
+import numpy.typing as npt
+
+from zarr.common import ChunkCoords, product
 
 if TYPE_CHECKING:
+    from zarr.array import Array
+    from zarr.buffer import NDArrayLike
     from zarr.chunk_grids import ChunkGrid
 
+BasicSelector = int | slice | EllipsisType
+BasicSelectorTuple = tuple[BasicSelector, ...]
+BasicSelection = BasicSelector | BasicSelectorTuple
+BasicSelectionNormalized = tuple[int | slice, ...]
+CoordinateSelection = npt.NDArray[np.intp]
+BlockSelector = int | slice
+BlockSelection = BlockSelector | tuple[BlockSelector, ...]
+BlockSelectionNormalized = tuple[BlockSelector, ...]
+MaskSelection = npt.NDArray[np.bool_]
+OrthogonalSelector = int | slice | npt.NDArray[np.intp | np.bool_]
```

Comment:
we should look for a way to simplify this as much as we can.`OrthogonalSelector` vs `OrthogonalSelection`, where the latter contains the former, is a recipe for confusion.